nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java

```diff
@@ -0,0 +1,257 @@
+
+  // Set to true when the current visited EnsuresNonNullIf method has correct semantics
+  private boolean semanticsHold;
+  // List of fields missing in the current EnsuresNonNullIf method so we can build proper error
+  // message
+  @Nullable private Set<String> missingFieldNames;
+
+  public EnsuresNonNullIfHandler() {
+    super("EnsuresNonNullIf");
+  }
+
+  /**
+   * Validates whether all parameters mentioned in the @EnsuresNonNullIf annotation are guaranteed
+   * to be {@code @NonNull} at exit point of this method.
+   */
+  @Override
+  protected boolean validateAnnotationSemantics(
+      MethodTree tree, MethodAnalysisContext methodAnalysisContext) {
+    // If no body in the method, return false right away
+    // TODO: Do we need proper error message here?
+    if (tree.getBody() == null) {
+      return false;
+    }
+
+    // clean up state variables, as we are visiting a new annotated method
+    semanticsHold = false;
+    missingFieldNames = null;
+
+    // We force the nullness analysis of the method
+    NullAway analysis = methodAnalysisContext.analysis();
+    VisitorState state = methodAnalysisContext.state();
+    analysis
+        .getNullnessAnalysis(state)
+        .forceRunOnMethod(new TreePath(state.getPath(), tree), state.context);
+
+    // If listed fields aren't checked for their nullness, return error
+    if (!semanticsHold) {
+      String message;
+      if (missingFieldNames == null) {
+        message = "Method is annotated with @EnsuresNonNullIf but does not return boolean";
+      } else {
+        message =
+            String.format(
+                "Method is annotated with @EnsuresNonNullIf but does not ensure fields %s",
+                missingFieldNames);
+      }
+      state.reportMatch(
+          analysis
+              .getErrorBuilder()
+              .createErrorDescription(
+                  new ErrorMessage(ErrorMessage.MessageTypes.POSTCONDITION_NOT_SATISFIED, message),
+                  tree,
+                  analysis.buildDescription(tree),
+                  state,
+                  null));
+    }
+
+    return true;
+  }
+
+  /*
+   * Sub-classes can only strengthen the post-condition. We check if the list in the child classes
+   * is at least the same as in the parent class.
+   */
+  @Override
+  protected void validateOverridingRules(
+      Set<String> overridingFieldNames,
+      NullAway analysis,
+      VisitorState state,
+      MethodTree tree,
+      Symbol.MethodSymbol overriddenMethod) {
+    FieldContractUtils.validateOverridingRules(
+        annotName, overridingFieldNames, analysis, state, tree, overriddenMethod);
+  }
+
+  @Override
+  public void onDataflowVisitReturn(
+      ReturnTree tree, NullnessStore thenStore, NullnessStore elseStore) {
+    if (visitingAnnotatedMethod) {
+      // We might have already found another return tree that results in what we need,
+      // so we don't keep going deep.
+      if (semanticsHold) {
```

Comment:
Any thoughts about this? A method with multiple returns, can we be smart about it?

Reply:
See comment above about checking `return` statements individually